models/learner.py

```diff
@@ -247,333 +275,465 @@ def loadSavedModel(log_folder, valid_models, cuda=True):
-                if self.use_triplets:
-                    tripletLoss(states, positive_states, negative_states, weight=self.losses_weights_dict['triplet'],
-                                loss_manager=loss_manager, alpha=0.2)
-                # Compute weighted average of losses
-                loss_manager.updateLossHistory()
-                loss = loss_manager.computeTotalLoss()
-
-                # We have to call backward in both train/val
-                # to avoid memory error
-                loss.backward()
-                if validation_mode:
-                    val_loss += loss.item()
-                    # We do not optimize on validation data
-                    # so optimizer.step() is not called
+            for valid_mode, dataloader in enumerate([dataloader_train, dataloader_valid]): 
+                if monitor_mode == 'pbar':
+                    pbar = tqdm(total=len(dataloader))
+                if self.model_type == 'gan':
+                    # GAN's training requires multi-optimizers, thus multiple loss_manager/epoch_loss/val_loss, etc.
+                    epoch_loss_D, epoch_batches_D = 0, 0
+                    epoch_loss_G, epoch_batches_G = 0, 0
+                    d_acc, g_acc = 0, 0
+                epoch_loss, epoch_batches = 0, 0
+                n_batch_per_epoch = len(dataloader)
+
+                start_time = time.time()
+                if valid_mode:
+                    self.module.eval()
+                    # with torch.no_grad():
+                    self.prev_grad_mode = torch.is_grad_enabled()
+                    torch.set_grad_enabled(False)
+                    # torch._C.set_grad_enabled(False) ??
                 else:
-                    self.optimizer.step()
-                    epoch_loss += loss.item()
-                    epoch_batches += 1
-                pbar.update(1)
-            pbar.close()
-
-            train_loss = epoch_loss / float(epoch_batches)
-            val_loss /= float(n_val_batches)
+                    self.module.train()
+                dataloader = infinite_dataloader(dataloader)#iter(dataloader)
+                for iter_ind in range(n_batch_per_epoch):
+                    (sample_idx, obs, next_obs, action, reward, noisy_obs, next_noisy_obs) = next(dataloader)
+                    obs, next_obs = obs.to(self.device), next_obs.to(self.device)
+                
+                    # if self.use_dae:
+                    #     noisy_obs = noisy_obs.to(self.device)
+                    #     next_noisy_obs = next_noisy_obs.to(self.device)    
+                        
+                    if self.model_type != "gan":
+                        ## It's extremely important to release loss tensor, otherwise it will cause 'memory leak".
+                        self.optimizer.zero_grad()
+                        loss_manager.resetLosses() 
+
+                    # Predict states given observations as in Time Contrastive Network (Triplet Loss) [Sermanet et al.]
```

Comment:
Here as well, it seems that some loses/ models are not taken into account yet...